### PR TITLE
Change Log in to use Fxapom for Desktop and Android

### PR DESCRIPTION
**Pull request by AndreiH**
I am trying to find the best and simplest way possible to use Fxapom in our Marketplace tests.
Of course Fxapom is at an early stage and it needs more updates to have something stable like Bidpom.
Tests worked fine for me locally, both Mobile and Desktop.

Here are the two adhoc :
http://selenium.qa.mtv2.mozilla.com:8080/view/Marketplace/job/marketplace.DEV.saucelabs.fxapom/
http://selenium.qa.mtv2.mozilla.com:8080/view/Marketplace/job/marketplace.DEV.mobile.saucelabs.fxapom/

i would appreciate  some feedback :) Thanks!

### CODE REVIEW

File: pages/desktop/consumer_pages/base.py [hunk 1]
```diff
@@ -68,9 +71,15 @@ def set_region(self, region):
...
-        fxa.login_user(user)
+        credentials = isinstance(user, MockUser) and user or \
+                    self.testsetup.credentials.get(user, FxATestAccount(use_prod=False).create_account())
+        from fxapom.pages.desktop.sign_in import SignIn
+        fxa_login = SignIn(self.testsetup)
+        if user=='default':
+            fxa_login.sign_in(credentials['email'], credentials['password'])
+        elif user=='new':
+            fxa_login.sign_in(credentials.email, credentials.password)
+
```
Comment (m8ttyB) [inline pages/desktop/consumer_pages/base.py]: Does this need need to take into account user credentials other than `default` and `new`? 
Also, should `user=None` be set to something more sensible in `def login(self, user=None):`

Comment (AndreiH) [inline pages/desktop/consumer_pages/base.py]: I tried to do something simple here. Credentials files doesn't need any changes. If user is specified as <code>"default"</code> in the tests then use the default user, else use the one that has just been created with FxATestAccount

Comment (m8ttyB) [inline pages/desktop/consumer_pages/base.py]: If that's the case -- note I don't have domain knowledge for this project -- I would suggest changing `user=None` to `user=default` and then adding an `else` conditional that returns an exception.

Comment (AndreiH) [inline pages/desktop/consumer_pages/base.py]: That should simplify things :+1:

Comment (bobsilverberg) [inline pages/desktop/consumer_pages/base.py]: I'm sorry, but this approach really rubs me the wrong way. I don't think that a `login` method should be responsible for creating a test user account. I think the account should be created before calling the login method and then the user can be passed in. Otherwise it's not going to be clear from the test that we are creating a new account - one would have to know the internals of the `login` method to know this.

Also, if we do that we can likely simplify the code in here, which is a bit hard to follow.

@AndreiH and @m8ttyB what do you think about this?

Comment (m8ttyB) [inline pages/desktop/consumer_pages/base.py]: @bobsilverberg I'm a +1 for reasoning .. I wasn't sure if @AndreiH's solution was a route that had already been agreed upon so I didn't comment on it.

File: pages/mobile/account_settings.py [hunk 2]
```diff
@@ -41,13 +43,22 @@ def click_logout(self):
         return Home(self.testsetup)
 
     def login(self, user=None):
-        fxa = self.click_sign_in()
-        fxa.login_user(user)
+        self.credentials = isinstance(user, MockUser) and user or \
```
Comment (bobsilverberg) [inline pages/mobile/account_settings.py]: Same comment as above here. I really don't think that `login` should create a user.

File: pages/mobile/details.py [hunk 3]
```diff
@@ -68,10 +69,11 @@ def click_view_reviews(self):
         from pages.mobile.reviews import Reviews
         return Reviews(self.testsetup)
 
-    def login_with_user_from_other_pages(self, user="default"):
-        from pages.fxa import FirefoxAccounts
-        fxa = FirefoxAccounts(self.testsetup)
-        fxa.login_user(user)
+    def login_with_user_from_other_pages(self):
+        credentials = FxATestAccount(use_prod=False).create_account()
```
Comment (bobsilverberg) [inline pages/mobile/details.py]: This is confusing too. The method is called `login_with_user_from_other_pages`, and this one *always* creates a new user account, whereas `login` has the option of passing in an existing user.

File: tests/desktop/consumer_pages/test_details_page.py [hunk 4]
```diff
@@ -95,7 +95,7 @@ def test_that_reports_abuse_as_signed_in_user(self, mozwebqa):
 
         Assert.true(home_page.is_the_current_page)
         home_page.header.click_sign_in()
-        home_page.login(user="default")
+        home_page.login(user="new")
```
Comment (bobsilverberg) [inline tests/desktop/consumer_pages/test_details_page.py]: Oh, I see here you're passing in `new` to indicate it's a new user, but that is still kind of confusing, as, theoretically, there could be a `new` entry in the credentials file. Also, the `login` method doesn't specifically look for a `new` user, so you could really pass anything here other than `default`. That seems like a loose API.

File: pages/mobile/details.py (hunk 3 above)
Comment (m8ttyB) [inline pages/mobile/details.py]: +1

File: tests/desktop/consumer_pages/test_details_page.py (hunk 4 above)
Comment (m8ttyB) [inline tests/desktop/consumer_pages/test_details_page.py]: +1 You might consider creating an api closer to what we've done on other projects, specifically creating a new user and then logging in with it -- much cleaner and more readable.

Comment (AndreiH) [inline tests/desktop/consumer_pages/test_details_page.py]: Thanks alot @bobsilverberg and @m8ttyB for the reviews.  I wasn't so pleased with this type of implementing the login. More views on how to implement this are needed. 
So, I simplified the login method, also adding to Base <code>create_new_user</code> and <code>get_default_user</code> methods that should return the correct user.

**Comment by AndreiH** [comment]
The tests work great for me locally, but I get intermittent failures on the <code>__init__</code> from Sign in, and I am seeing this also in the adhocs. I resolved this by adding a <code>time.sleep</code> here [1], and we should add a better wait. 
Also I am seeing Mobile failures when the email or password is entered incorrectly [2], but I couldn't reproduce this locally, so I think it's because a known bug in Chrome Driver, see [3].

[1] https://github.com/bobsilverberg/fxapom/blob/remove_fxos/fxapom/pages/desktop/sign_in.py#L29
[2] https://saucelabs.com/jobs/3ffbdd64195b4498b5a3163379f90acd
[3] https://github.com/mozilla/marketplace-tests/issues/568

**Comment by bobsilverberg** [comment]
@AndreiH I fixed the issue with fxapom and have pushed a new version to my `master` branch, so you should update the `requirements.txt` to point to my master branch which should address the timeout issue.

File: pages/desktop/consumer_pages/base.py [hunk 5]
```diff
@@ -67,13 +70,22 @@ def set_region(self, region):
 
         self.wait_notification_box_not_visible()
 
-    def login(self, user=None):
-        from pages.fxa import FirefoxAccounts
-        fxa = FirefoxAccounts(self.testsetup)
-        fxa.login_user(user)
+    def login(self, email, password):
```
Comment (bobsilverberg) [inline pages/desktop/consumer_pages/base.py]: Thanks for making this change, @AndreiH. I think that the `login` method could accept a user object, rather than the `email` and `password. You have set up  `create_new_user` and `get_default_user` to both return a `MockUser` object, so it would be fine for `login` to expect this object. That would mean that the tests wouldn't have to extract the `email` and `password` from the user.

File: pages/desktop/consumer_pages/base.py [hunk 6]
```diff
@@ -67,13 +70,22 @@ def set_region(self, region):
...
+        fxa_login.sign_in(email, password)
+
         self.wait_notification_box_visible()
         self.wait_notification_box_not_visible()
 
+    def create_new_user(self):
+        acct = FxATestAccount(use_prod=False).create_account()
+        return MockUser(email=acct.email, password=acct.password, name=acct.email.split('@')[0])
+
+    def get_default_user(self, user='default'):
```
Comment (bobsilverberg) [inline pages/desktop/consumer_pages/base.py]: This seems a bit odd that the method is called `get_default_user`, and yet it accepts an argument for `user` which defaults to `default`. That would mean I could call `get_default_user('bob')` and it would return the credentials for `bob`, which seems odd because the name of the method is `get_default_user`. Maybe it should be renamed to simply `get_user`.

File: pages/mobile/base.py [hunk 7]
```diff
@@ -64,6 +64,19 @@ def wait_notification_box_visible(self):
     def wait_notification_box_not_visible(self):
         self.wait_for_element_not_visible(*self._notification_locator)
 
+    def login(self, email, password):
```
Comment (bobsilverberg) [inline pages/mobile/base.py]: Same comment as above about `login` with `email` and `password`.

File: pages/mobile/base.py [hunk 8]
```diff
@@ -64,6 +64,19 @@ def wait_notification_box_visible(self):
...
+        from fxapom.pages.desktop.sign_in import SignIn
+        fxa_login = SignIn(self.testsetup)
+        fxa_login.sign_in(email, password)
+
+    def create_new_user(self):
+        acct = FxATestAccount(use_prod=False).create_account()
+        return MockUser(email=acct.email, password=acct.password, name=acct.email.split('@')[0])
+
+    def get_default_user(self, user='default'):
```
Comment (bobsilverberg) [inline pages/mobile/base.py]: Same as above, suggest renaming to `get_user`.

File: pages/mobile/base.py [hunk 9]
```diff
@@ -64,6 +64,19 @@ def wait_notification_box_visible(self):
...
+        acct = self.testsetup.credentials[user]
+        return MockUser(email=acct['email'], password=acct['password'], name=acct['email'].split('@')[0])
+
```
Comment (bobsilverberg) [inline pages/mobile/base.py]: Actually, these look pretty much identical to the same named methods in `pages/desktop/consumer_pages/base.py`. I would consider creating a higher level object that both `base.py`s could inherit from, which would avoid this duplication.

File: requirements.txt [hunk 10]
```diff
@@ -13,3 +13,4 @@ chardet==2.1.1
 execnet==1.1
 requests==2.4.3
 Marketplace==0.9.1
+-e git+https://github.com/bobsilverberg/fxapom.git@remove_fxos#egg=fxapom
```
Comment (bobsilverberg) [inline requirements.txt]: As mentioned, please update to point at the `master` branch.

File: tests/desktop/consumer_pages/test_users_account.py [hunk 11]
```diff
@@ -35,7 +38,8 @@ def test_user_can_sign_in_and_sign_out_from_home_page(self, mozwebqa):
         home_page.go_to_homepage()
 
         home_page.header.click_sign_in()
-        home_page.login(user="default")
+        acct = home_page.get_default_user()
```
Comment (bobsilverberg) [inline tests/desktop/consumer_pages/test_users_account.py]: Is there a reason that the tests in this file are using the default user instead of creating a new user? It seems to me that it would be safer to always create a new user, unless there is a specific reason for needing an existing user for a test.

File: tests/desktop/consumer_pages/test_users_account.py [hunk 12]
```diff
@@ -100,11 +109,6 @@ def test_editing_user_profile(self, mozwebqa):
         profile_page.refresh_page()
         Assert.equal(profile_page.display_name, name)
 
-        # Undo the changes
-        profile_page.edit_display_name(initial_value)
-        profile_page.save_changes()
-        profile_page.wait_notification_box_not_visible()
-
     @pytest.mark.credentials
     def test_that_checks_changing_language(self, mozwebqa):
```
Comment (bobsilverberg) [inline tests/desktop/consumer_pages/test_users_account.py]: I know it isn't related to this PR, but is there a reason we are keeping this test around? Will changing languages be implemented in fireplace at some point in the future? @krupa what do you think?

**Comment by bobsilverberg** [comment]
@AndreiH Thanks for the update. The tests are all passing for me on desktop now, but I have made some comments for improvements. Nearly all of the tests are failing on mobile, most of them here:

```
self = <pages.mobile.home.Home object at 0x10aa42250>

    @property
    def first_app_name(self):
>       return self.find_element(*self._first_app_name_locator).text
```

with selenium unable to find that element.

**Comment by bobsilverberg** [comment]
This seems to be (one of) the problems:

0002screenshot

File: tests/desktop/consumer_pages/test_users_account.py (hunk 11 above)
Comment (AndreiH) [inline tests/desktop/consumer_pages/test_users_account.py]: I updated to create a new user.

File: requirements.txt (hunk 10 above)
Comment (AndreiH) [inline requirements.txt]: done

File: pages/mobile/base.py (hunk 9 above)
Comment (AndreiH) [inline pages/mobile/base.py]: I moved both methods to base_test.py, and also moved base_test.py to Pages, because it was only located in Pages.desktop.

File: pages/mobile/base.py (hunk 8 above)
Comment (AndreiH) [inline pages/mobile/base.py]: done

File: pages/mobile/base.py (hunk 7 above)
Comment (AndreiH) [inline pages/mobile/base.py]: done

File: pages/desktop/consumer_pages/base.py (hunk 6 above)
Comment (AndreiH) [inline pages/desktop/consumer_pages/base.py]: done. Thanks!

File: pages/desktop/consumer_pages/base.py (hunk 5 above)
Comment (AndreiH) [inline pages/desktop/consumer_pages/base.py]: done. Thanks!

**Comment by AndreiH** [comment]
@bobsilverberg I updated the tests according reviews. I also updated Dev_Hub login as I removed fxa.py 
The Mobile failure from the screenshot that you posted is not reproducing anymore, but I am seeing an intermittent failure in <code>login</code>  method, see [1], that I could not reproduce locally.
There are only 7 characters entered in the password field.
This might be because: https://github.com/mozilla/marketplace-tests/issues/568

[1] https://saucelabs.com/tests/284fe8442733448c9df27304506ef8b4

File: pages/desktop/consumer_pages/base.py [hunk 13]
```diff
@@ -67,10 +68,12 @@ def set_region(self, region):
 
         self.wait_notification_box_not_visible()
 
-    def login(self, user=None):
-        from pages.fxa import FirefoxAccounts
-        fxa = FirefoxAccounts(self.testsetup)
-        fxa.login_user(user)
+    def login(self, user):
+        credentials = isinstance(user, MockUser) and user
```
Comment (bobsilverberg) [inline pages/desktop/consumer_pages/base.py]: This does not really make sense. If `isinstance(user, MockUser)` returns `False`, then `credentials` will also be `False`, which will cause an exception at line 75 below. In fact, I am seeing exactly that in 3 of the tests when I run them locally. See below for one example:

```
________________________________________________ TestConsumerPage.test_footer_section_links _________________________________________________

self = <tests.desktop.consumer_pages.test_home_page.TestConsumerPage instance at 0x110718a28>
mozwebqa = <pytest_mozwebqa.pytest_mozwebqa.TestSetup instance at 0x110718638>

    @pytest.mark.sanity
    @pytest.mark.nondestructive
    def test_footer_section_links(self, mozwebqa):
        home_page = Home(mozwebqa)
        home_page.go_to_homepage()

        home_page.header.click_sign_in()
>       home_page.login(user="default")

tests/desktop/consumer_pages/test_home_page.py:142:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pages.desktop.consumer_pages.home.Home object at 0x110711350>, user = 'default'

    def login(self, user):
        credentials = isinstance(user, MockUser) and user
        from fxapom.pages.desktop.sign_in import SignIn
        fxa_login = SignIn(self.testsetup)
>       fxa_login.sign_in(credentials['email'], credentials['password'])
E       TypeError: 'bool' object has no attribute '__getitem__'

pages/desktop/consumer_pages/base.py:78: TypeError
```

File: pages/desktop/developer_hub/base.py [hunk 14]
```diff
@@ -8,14 +8,17 @@
...
+from tests.base_test import BaseTest
 from pages.page import Page
 
 
 class Base(Page):
 
-    def login(self, user="default"):
-        fxa = self.header.click_login()
-        fxa.login_user(user)
+    def login(self, mozwebqa, user="default"):
```
Comment (bobsilverberg) [inline pages/desktop/developer_hub/base.py]: I find this confusing and inconsistent. For consumer pages we changed the `login` method to just accept a user object. I think this would be cleaner and simpler if we did the same thing here.

File: pages/mobile/base.py [hunk 15]
```diff
@@ -64,6 +64,12 @@ def wait_notification_box_visible(self):
     def wait_notification_box_not_visible(self):
         self.wait_for_element_not_visible(*self._notification_locator)
 
+    def login(self, user):
+        credentials = isinstance(user, MockUser) and user
```
Comment (bobsilverberg) [inline pages/mobile/base.py]: The same problem as above exists here. What happens if `isinstance(user, MockUser)` returns `False`? If that is a possibility then you need a code path to deal with that.

File: tests/base_test.py [hunk 16]
```diff
@@ -75,6 +76,13 @@ def _take_first_new_app_name(self, mozwebqa):
         home_page = Home(mozwebqa)
         home_page.go_to_homepage()
 
-        home_page.click_new_tab()
         app_name = home_page.first_new_app_name
         return app_name
+
+    def _create_new_user(self):
```
Comment (bobsilverberg) [inline tests/base_test.py]: I'm not sure that we need to prefix these with `_`. They are methods in a `BaseTest` object, so I wouldn't expect anyone to try to call them from elsewhere. I think the `_` is redundant, and it's also not something we usually do in these objects.

**Comment by bobsilverberg** [comment]
@AndreiH Thanks for the update. There are a few things that still need to be addressed, plus this is going to need a rebase as two other PRs have been merged. I would suggest squashing all commits to date right now, and resolving any merge conflicts, to make it a bit cleaner, and then adding a new commit to address the most recent comments.

Also, 3 of the mobile tests are failing for me on Saucelabs, and it doesn't look like the entry issue you mentioned. Here's one stack trace:

```
________________________________________________ TestAccounts.test_user_can_login_and_logout ________________________________________________
[gw1] darwin -- Python 2.7.6 /Users/bsilverberg/.virtualenvs/marketplace-tests/bin/python
self = <tests.mobile.test_users_account.TestAccounts instance at 0x10577db48>
mozwebqa = <pytest_mozwebqa.pytest_mozwebqa.TestSetup instance at 0x10577d998>

    @pytest.mark.nondestructive
    def test_user_can_login_and_logout(self, mozwebqa):

        home_page = Home(mozwebqa)
        home_page.go_to_homepage()

        settings_page = home_page.header.click_settings()
        settings_page.click_sign_in()

        acct = self._create_new_user()
>       settings_page.login(acct)

tests/mobile/test_users_account.py:26:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pages.mobile.account_settings.AccountSettings object at 0x10577b150>
user = {'name': 'test-cee5b76da8', 'password': '15c61bf3', 'email': 'test-cee5b76da8@restmail.net'}

    def login(self, user):
        credentials = isinstance(user, MockUser) and user
        from fxapom.pages.desktop.sign_in import SignIn
        fxa_login = SignIn(self.testsetup)
>       fxa_login.sign_in(credentials['email'], credentials['password'])

pages/mobile/base.py:71:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fxapom.pages.desktop.sign_in.SignIn object at 0x1040612d0>
email = 'test-cee5b76da8@restmail.net', password = '15c61bf3'

    def sign_in(self, email, password):
        """Signs in using the specified email address and password."""
        self.email = email
        if self.is_element_present(*self._next_button_locator):
            self.click_next(expect='password')
        self.login_password = password
>       self.click_sign_in()

../../.virtualenvs/marketplace-tests/src/fxapom/fxapom/pages/desktop/sign_in.py:92:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fxapom.pages.desktop.sign_in.SignIn object at 0x1040612d0>

    def click_sign_in(self):
        self.selenium.find_element(*self._sign_in_locator).click()
        if self.popup:
            WebDriverWait(self.selenium, self.timeout).until(
>                       lambda s: self._sign_in_window_handle not in self.selenium.window_handles)

../../.virtualenvs/marketplace-tests/src/fxapom/fxapom/pages/desktop/sign_in.py:72:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <selenium.webdriver.support.wait.WebDriverWait object at 0x1044dfad0>
method = <function <lambda> at 0x1057830c8>, message = ''

    def until(self, method, message=''):
        """Calls the method provided with the driver as an argument until the \
            return value is not False."""
        end_time = time.time() + self._timeout
        while(True):
            try:
                value = method(self._driver)
                if value:
                    return value
            except self._ignored_exceptions:
                pass
            time.sleep(self._poll)
            if(time.time() > end_time):
                break
>       raise TimeoutException(message)
E       TimeoutException: Message:

../../.virtualenvs/marketplace-tests/lib/python2.7/site-packages/selenium/webdriver/support/wait.py:71: TimeoutException
-------------------------------------------------------------- pytest-mozwebqa --------------------------------------------------------------
```

**Comment by AndreiH** [comment]
So, as discussed on IRC, the failures that we see on login are caused by https://support.saucelabs.com/requests/16244.
We should update Fxapom to return a password without digits, a workaround suggested also by Saucelabs Support

File: pages/desktop/consumer_pages/base.py (hunk 13 above)
Comment (AndreiH) [inline pages/desktop/consumer_pages/base.py]: Changed

File: pages/desktop/developer_hub/base.py (hunk 14 above)
Comment (AndreiH) [inline pages/desktop/developer_hub/base.py]: updated

File: pages/mobile/base.py (hunk 15 above)
Comment (AndreiH) [inline pages/mobile/base.py]: removed isinstance

**Comment by AndreiH** [comment]
@bobsilverberg another update of this PR. Bear with me ! :)

**Comment by AndreiH** [comment]
@bobsilverberg Desktop and Mobile tests are updated and they are working great.
I am seeing https://github.com/mozilla/marketplace-tests/issues/568 affecting when logging in with a new user, digits are not correctly inserted for user.
See: https://saucelabs.com/tests/2d1b85eea8614f5dbc89164c3e1709d3

**Comment by bobsilverberg** [comment]
I merged the changes to fxapom to force the email address to have no numbers. The adhoc mobile job seems to be failing with a different issue, so I think we should merge this, and then address that issue separately if it persists. I will squash it for you to get it in today. Thanks for all your hard work, @AndreiH!

**Comment by bobsilverberg** [comment]
Actually, @AndreiH, I see you've already pulled changes in a couple of times, so there are conflicts. I will leave it to you to squash the commits tomorrow, then we can merge.

**Comment by AndreiH** [comment]
@bobsilverberg I squashed the commits and also solved the mobile failures.
Instead of review uniqueness with current time I added a random string.
Please see https://github.com/AndreiH/marketplace-tests/commit/3cc91fbb75cf2f4d475df1d8c9d7247e3494a23c

**Comment by AndreiH** [comment]
I just realized that creating an account on Stage also is using the Prod API url <code>self.fxa_url = 'https://api.accounts.firefox.com/'</code>, so that means we can't use variable <code>use_prod=False</code> for creating an account. We will have to do this by using the <code>base_url</code> condition : https://github.com/bobsilverberg/fxapom/blob/master/fxapom/fxapom.py#L33

**Comment by bobsilverberg** [comment]
Looks great @AndreiH, and everything passes! Can you please squash and then we'll merge?

**Comment by bobsilverberg** [comment]
Perfect! Thanks @AndreiH! :cake:
